src/main/java/uk/gov/hmcts/ccd/domain/service/casedataaccesscontrol/AccessProfileServiceImpl.java

```diff
@@ -0,0 +1,81 @@
+package uk.gov.hmcts.ccd.domain.service.casedataaccesscontrol;
+
+import org.apache.commons.collections.CollectionUtils;
+import org.springframework.stereotype.Component;
+import uk.gov.hmcts.ccd.domain.model.casedataaccesscontrol.AccessProfile;
+import uk.gov.hmcts.ccd.domain.model.casedataaccesscontrol.RoleAssignment;
+import uk.gov.hmcts.ccd.domain.model.casedataaccesscontrol.RoleAssignmentFilteringResult;
+import uk.gov.hmcts.ccd.domain.model.definition.RoleToAccessProfileDefinition;
+import uk.gov.hmcts.ccd.domain.service.AccessControl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+@Component
+public class AccessProfileServiceImpl implements AccessProfileService, AccessControl {
+
+    @Override
+    public List<AccessProfile> generateAccessProfiles(RoleAssignmentFilteringResult filteringResults,
+                                                      List<RoleToAccessProfileDefinition> roleToAccessProfiles) {
+
+        List<AccessProfile> accessProfiles = new ArrayList<>();
+        Map<String, RoleToAccessProfileDefinition> roleToAccessProfileDefinitionMap =
+            toRoleNameAsKeyMap(roleToAccessProfiles);
+
+        for (RoleAssignment roleAssignment : filteringResults.getRoleAssignments()) {
+
+            RoleToAccessProfileDefinition roleToAccessProfileDefinition =
+                roleToAccessProfileDefinitionMap.get(roleAssignment.getRoleName());
+
+            if (roleToAccessProfileDefinition != null && !roleToAccessProfileDefinition.isDisabled()) {
+                List<String> authorisations = roleToAccessProfileDefinition.getAuthorisationList();
+                List<String> roleAssignmentAuthorisations = roleAssignment.getAuthorisations();
+
+                if (authorisationsAllowMappingToAccessProfiles(authorisations, roleAssignmentAuthorisations)) {
+                    accessProfiles.addAll(createAccessProfiles(roleAssignment, roleToAccessProfileDefinition));
+                }
+            }
+        }
+        return accessProfiles;
```

Comment:
Is it possible (and ok) that we could end up with duplicates in this list?

Reply:
if we endup with duplicates then that is the problem from service sending those access profiles. but at then end we are extracting the set of names, that should be fine